src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,715 @@
+/*********************                                                        */
+/*! \file bvgauss.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Aina Niemetz
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Gaussian Elimination preprocessing pass.
+ **
+ ** Simplify a given equation system modulo a (prime) number via Gaussian
+ ** Elimination if possible.
+ **/
+
+#include "theory/bv/bvgauss.h"
+
+#include <stack>
+#include <unordered_map>
+
+#include "theory/rewriter.h"
+#include "theory/bv/theory_bv_utils.h"
+#include "theory/bv/theory_bv_rewrite_rules_normalization.h"
+
+using namespace CVC4;
+using namespace std;
+
+namespace CVC4 {
+namespace theory {
+namespace bv {
+
+static bool is_bv_const(Node n)
+{
+  if (n.isConst()) { return true; }
+  return Rewriter::rewrite(n).getKind() == kind::CONST_BITVECTOR;
+}
+
+static Node get_bv_const(Node n)
+{
+  Assert(is_bv_const(n));
+  return Rewriter::rewrite(n);
+}
+
+static Integer get_bv_const_value(Node n)
+{
+  Assert(is_bv_const(n));
+  return get_bv_const(n).getConst<BitVector>().getValue();
+}
+
+/* Note: getMinBwExpr assumes that 'expr' is rewritten.
+ *
+ * If not, all operators that are removed via rewriting (e.g., ror, rol, ...)
+ * will be handled via the default case, which is not incorrect but also not
+ * necessarily the minimum. */
+unsigned BVGaussElim::getMinBwExpr(Node expr)
+{
+  stack<Node> visit;
+  /* Maps visited nodes to the determined minimum bit-width required. */
+  unordered_map<Node, unsigned, NodeHashFunction> visited;
+  unordered_map<Node, unsigned, NodeHashFunction>::iterator it;
+
+  /* Rewrite const expr, overflows in consts are irrelevant. */
+  if (is_bv_const(expr)) { expr = get_bv_const(expr); }
+
+  visit.push(expr);
+  while (!visit.empty())
+  {
+    Node n = visit.top();
+    visit.pop();
+    it = visited.find(n);
+    if (it == visited.end())
+    {
+      visited[n] = 0;
+      visit.push(n);
+      for (const Node &nn : n)
+      {
+        visit.push(nn);
+      }
+    }
+    else if (it->second == 0)
+    {
+      /* Rewrite const expr, overflows in consts are irrelevant. */
+      if (is_bv_const(n)) { n = get_bv_const(n); }
```

Comment:
Actually, it seems like it would be slightly more efficient to put this check after line 74 (i.e. have something like `if (is_bv_const(n)) { visited[n] = get_bv_const(n).length(); }`) to avoid visiting the children. Also, the check on line 65 would then be obsolete.

Reply:
I agree that it is possible to shortcut the traversal here.

However, this does not make the check on line 65 obsolete. Line 65 rewrites `expr` in case it is a const expression since it won't be in `visited` otherwise (only the rewritten expression will be).

Reply:
I will keep it as it is since it is slightly more readable and does not make any difference in terms of performance.

Reply:
Hmm, let's assume that `expr` rewrites to a const, you removed the code on line 65, and made the changes that I proposed. The unrewritten `expr` is in `visit`, but not in `visited` and it rewrites to const, so we would do `visited[n] = get_bv_const(n).length();` where `n` is `expr` and add no children. Since `visit` is empty, we would return `get_bv_const(n).length()` as the length, which seems to be what we expect. I am also not sure if the condition on line 196 is needed (and it also seems a bit dubious with the assertion on line 195). It seems to me like the easiest solution would be to say that `visited` maps _unrewritten_ nodes to their length.

Reply:
Hm, if you map the unrewritten node to the resulting bit-width, line 65 becomes indeed obsolete (previously, for const expr the rewritten node was mapped). Consequently, it may then happen that const expressions that map to the same value are mapped not only once but in as many variations as they appear (e.g. `(_ bv1 4)` = `((_ extract 3 0) (_ bv1 10))` would result in two mappings rather than one). Which probably doesn't matter too much in this case.

Line 196 is a leftover from a refactor and in particular obsolete with the changes above. Will fix.